net/graphsync_fetcher_test.go

```diff
@@ -231,19 +310,224 @@ func TestGraphsyncFetcher(t *testing.T) {
+		mgs.expectRequestToRespondWithLoader(pid0, recursiveSelector(4), errorInMultiBlockLoader, penultimate.At(0).Cid())
+
+		fetcher := net.NewGraphSyncFetcher(ctx, mgs, bs, bv, mv, &fakePeerTracker{})
+		done := doneAt(gen.Key())
+
+		ts, err := fetcher.FetchTipSets(ctx, final.Key(), pid0, done)
+		mgs.verifyReceivedRequestCount(3)
+		mgs.verifyExpectations()
+		require.Errorf(t, err, "Failed fetching tipset: %s", multi.Key().String())
+		require.Nil(t, ts)
+	})
+
+	t.Run("missing single block in multi block tip during recursive fetch, recover through fallback", func(t *testing.T) {
+		gen := builder.NewGenesis()
+		multi := builder.BuildOn(gen, 3, withMessageEachBuilder)
+		withMultiParent := builder.BuildOn(multi, 1, withMessageEachBuilder)
+		penultimate := builder.BuildManyOn(2, withMultiParent, withMessageBuilder)
+		final := builder.BuildOn(penultimate, 1, withMessageEachBuilder)
+		height, err := final.Height()
+		require.NoError(t, err)
+		chain1 := types.NewChainInfo(pid1, final.Key(), height)
+		chain2 := types.NewChainInfo(pid2, final.Key(), height)
+		pt := &fakePeerTracker{[]*types.ChainInfo{chain1, chain2}}
+
+		mgs := newMockableGraphsync(ctx, bs, t)
+		errorInMultiBlockLoader := errorOnCidsLoader(loader, multi.At(1).Cid())
+		mgs.expectRequestToRespondWithLoader(pid0, layer1Selector, errorInMultiBlockLoader, final.At(0).Cid())
+		mgs.expectRequestToRespondWithLoader(pid0, recursiveSelector(1), errorInMultiBlockLoader, final.At(0).Cid())
+		mgs.expectRequestToRespondWithLoader(pid0, recursiveSelector(4), errorInMultiBlockLoader, penultimate.At(0).Cid())
+		mgs.expectRequestToRespondWithLoader(pid1, recursiveSelector(4), loader, withMultiParent.At(0).Cid())
+
+		fetcher := net.NewGraphSyncFetcher(ctx, mgs, bs, bv, mv, pt)
+		done := doneAt(gen.Key())
+
+		ts, err := fetcher.FetchTipSets(ctx, final.Key(), pid0, done)
+		require.NoError(t, err, "the request completes successfully")
+		mgs.verifyReceivedRequestCount(4)
+		mgs.verifyExpectations()
+		require.Equal(t, 6, len(ts), "the right number of tipsets is returned")
+		expectedTs := final
+		for _, resultTs := range ts {
+			require.True(t, expectedTs.Key().Equals(resultTs.Key()), "the initial tipset is correct")
+			key, err := expectedTs.Parents()
+			require.NoError(t, err)
+			if !key.Empty() {
+				expectedTs, err = builder.GetTipSet(key)
+				require.NoError(t, err)
+			}
+		}
+	})
+
+	t.Run("stopping at edge heights in recursive fetch", func(t *testing.T) {
+		gen := builder.NewGenesis()
+		recursive16stop := builder.BuildManyOn(1, gen, withMessageBuilder)
+		recursive16middle := builder.BuildManyOn(15, recursive16stop, withMessageBuilder)
+		recursive4stop := builder.BuildManyOn(1, recursive16middle, withMessageBuilder)
+		recursive4middle := builder.BuildManyOn(3, recursive4stop, withMessageBuilder)
+		recursive1stop := builder.BuildManyOn(1, recursive4middle, withMessageBuilder)
+		final := builder.BuildOn(recursive1stop, 1, withMessageEachBuilder)
```

Comment:
The coupling of this test to the specific expansion strategy used by the fetcher worries me. It's very easy to imagine that code changing and this test becoming useless. Could we change it to loop and test every stopping point from 1 to 22?